docassemble/ALKilnInThePlayground/data/questions/run_alkiln_tests.yml

```diff
@@ -519,54 +511,103 @@ code: |
-        _TAGS=action_argument('tag_expression'),
-        # Don't need the values of these when running in interview
-        REPO_URL="X",
-        BRANCH_NAME="X",
-        DOCASSEMBLE_DEVELOPER_API_KEY="X",
-        **env_vars
-    ))
-    has_run_tests = True
-  
-  # https://github.com/SuffolkLITLab/docassemble-AssemblyLine/blob/main/docassemble/AssemblyLine/al_document.py#L1275-L1284
-  # https://github.com/SuffolkLITLab/docassemble-ALDashboard/blob/main/docassemble/ALDashboard/data/questions/compile_bootstrap.yml
-  
-  output = '<div class="console_output">\n'
-  output += '<div class="section">\n'
-  output += '<h2 id="report">Console output</h2>\n'
-  output += '<pre>\n<code>\n'
-  output += f'{test_output.stdout.decode("utf-8")}{test_output.stderr.decode("utf-8")}\n'
-  output += '</code>\n</pre>\n'
-  output += '</div>\n'
-  output += '</div>\n'
-  
-  log_output = f'ALKiln test run output:\n{test_output.stdout.decode("utf-8")}{test_output.stderr.decode("utf-8")}'
-  log(log_output)
-  
-  background_response(output)
+    try:
+      # Should we try/catch this function call?
+      process = subprocess.Popen(
+        to_run,
+        start_new_session=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        env=custom_env
+      )
+
+      # Ensure tests don't re-run, even (especially) if they error
+      has_run_tests = True
+
+      # Get the process id of the subprocess
+      test_run_pid = process.pid
+      log( f'□□□ Running ALKiln tests with process ID { test_run_pid }' )
+
+      start_process_wait_time = current_datetime()
+      # `.communicate()` `timeout` is in seconds
+      # 60 seconds * 60 minutes * 12 hours
+      max_run_time = env_vars.get(
+        'ALKILN_MAX_SECONDS_FOR_TEST_RUN',
+        60 * 60 * 12
+      )
+      # Get output tuple: `(stdout_data, stderr_data)`
+      # As a last resort, the processes will timeout
+      test_output = process.communicate( timeout=max_run_time )
+      returncode = process.returncode
+      
+    except Exception as error:
+      # We get here even when the sub-processes fail
+      log( f'□□□ ALKiln test subprocess completed with { returncode }' )
+      # If it's within 10 seconds of the max time, assume it's an
+      # error for taking too long. Log the max time error.
+      if date_difference( starting=start_process_wait_time ).seconds > (max_run_time - 10):
+        timeout_output = f'■■■ ALKiln Error ALKP0006: the tests run with the pid { test_run_pid } reached the maximum time allowed ({ max_run_time/24 } hours). You can change the maximum time by adding the ALKILN_MAX_SECONDS_FOR_TEST_RUN value to your config\'s `alkiln` key and giving it a different value.'
+        log(timeout_output)
+    
+      # Check if the process completed successfully
+      if returncode != 0:
+        # Reraise the error
+        raise
+    
+    finally:
+      # Always make sure subprocess and its children are terminated
+      try:
```

Comment:
The exception is `ProcessLookupError` which is a subprocess of `OSError`.

We could do `except OSError:` which should catch all likely failures from the os calls - but you've convinced me that we don't want to handle other causes of failure. Eg. if we don't have permission to kill it that is weird and shouldn't handle it.

What exceptions are likely to be raised by `process.wait` though?

Reply:
We discussed doing the same as a [comment above](https://github.com/SuffolkLITLab/docassemble-ALKilnInThePlayground/pull/39#discussion_r1495865321) - catch the specific error and reraise other errors.

Reply:
I don't know other exceptions. That's a good research question. We might only address it if it comes up as it will at least be handled with the current method, even if not in the most ideal way possible.